modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/recovery/RecoveryIoImp.java

```
@@ -0,0 +1,155 @@
+
+        this.log = log;
+        this.ctx = ctx;
+    }
+
+    /** {@inheritDoc} */
+    @Override public void send(String constId, Message msg) throws IgniteCheckedException {
+        UUID nodeId = idMapper.mapToUUID(constId);
+
+        assert nodeId != null;
+
+        ctx.io().sendToGridTopic(nodeId, TOPIC_RECOVERY, msg, SYSTEM_POOL);
+    }
+
+    /** {@inheritDoc} */
+    @Override public void receive(final IgniteBiInClosure<String, Message> handler) {
+        msgHandler = handler;
+    }
+
+    /** {@inheritDoc} */
+    @Override public void onNodeLeft(final IgniteInClosure<String> handler) {
+        nodeLeftHandler = handler;
+    }
+
+    /** {@inheritDoc} */
+    @Override public List<String> constIds(AffinityTopologyVersion ver) {
+        List<String> constIds = new ArrayList<>();
+
+        for (Object constId : idMapper.mapToConsistentId(ver))
+            constIds.add(constId.toString());
+
+        return constIds;
+    }
+
+    /** {@inheritDoc} */
+    @Override public String localNodeConsistentId() throws IgniteCheckedException {
+        try {
+            return ctx.pdsFolderResolver().resolveFolders().consistentId().toString();
+        }
+        catch (IgniteCheckedException e) {
+            U.error(log, "Fail resolve consistent id.", e);
+
+            new Thread(
+                new Runnable() {
+                    @Override public void run() {
+                        G.stop(true);
+                    }
+                }
+            ).start();
+
+            throw e;
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override public void onMessage(UUID nodeId, Object msg, byte plc) {
+        IgniteBiInClosure<String, Message> handler = msgHandler;
+
+        if (handler != null && msg instanceof Message) {
+            String constId = idMapper.mapToConsistentId(
+                ctx.discovery().topologyVersionEx(), nodeId
+            ).toString();
+
+            handler.apply(constId, (Message)msg);
+        }
```

We should not silently drop a message. If a listener is supposed to be always installed before a first message is sent, we should at least print a warning.